src/locator.rs

```diff
@@ -0,0 +1,643 @@
+            }),
+            Ok((org_id @ None, _)) => Ok(Locator {
+                fetcher,
+                org_id,
+                project,
+                revision,
+            }),
+            Err(error) => Err(Error::Parse(ParseError::Project {
+                input: locator.to_owned(),
+                project,
+                error,
+            })),
+        }
+    }
+
+    /// Promote a `Locator` to a [`StrictLocator`] by providing the default value to use
+    /// for the `revision` component, if one is not specified in the locator already.
+    ///
+    /// The `ToString` implementation is lazily evaluated if the locator doesn't already contain a revision.
+    pub fn promote_strict(self, revision: impl ToString) -> StrictLocator {
+        let locator = StrictLocator::builder()
+            .fetcher(self.fetcher)
+            .project(self.project)
+            .revision(self.revision.unwrap_or_else(|| revision.to_string()));
+
+        match self.org_id {
+            None => locator.build(),
+            Some(org_id) => locator.org_id(org_id).build(),
+        }
+    }
+
+    /// Promote a `Locator` to a [`StrictLocator`] by providing the default value to use
+    /// for the `revision` component, if one is not specified in the locator already.
+    ///
+    /// The revision is lazily evaluated if the locator doesn't already contain a revision.
+    pub fn promote_strict_with<F: Fn() -> String>(self, revision: F) -> StrictLocator {
+        let locator = StrictLocator::builder()
+            .fetcher(self.fetcher)
+            .project(self.project)
+            .revision(self.revision.unwrap_or_else(revision));
+
+        match self.org_id {
+            None => locator.build(),
+            Some(org_id) => locator.org_id(org_id).build(),
+        }
+    }
+
+    /// Converts the locator into a [`PackageLocator`] by discarding the `revision` component.
+    /// Equivalent to the `From` implementation, but offered as a method for convenience.
+    pub fn into_package(self) -> PackageLocator {
+        self.into()
+    }
+
+    /// Explodes the locator into its (owned) parts.
+    /// Used for conversions without cloning.
+    pub(crate) fn explode(self) -> (Fetcher, Option<usize>, String, Option<String>) {
+        (self.fetcher, self.org_id, self.project, self.revision)
+    }
+}
+
+impl Ord for Locator {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self.fetcher.cmp(&other.fetcher) {
+            Ordering::Equal => {}
+            ord => return ord,
+        }
+        match alphanumeric_sort::compare_str(&self.project, &other.project) {
+            Ordering::Equal => {}
+            ord => return ord,
+        }
+        match (&self.revision, &other.revision) {
+            (None, None) => Ordering::Equal,
+            (None, Some(_)) => Ordering::Greater,
+            (Some(_), None) => Ordering::Less,
+            (Some(a), Some(b)) => alphanumeric_sort::compare_str(a, b),
```

Comment:
This ordering is kind of a lie, because ordering of revisions is fetcher dependent, and I'm worried that someone will depend on this ordering function without knowing that.

We should at the very least add a comment here mentioning that.

Eventually we'll want to implement this per fetcher. For now, though, would it be worth using the `semver` crate and comparing them using that if both versions are valid semver?

Reply:
That's a good idea, I'll do that.

FWIW, I don't think we'll ever do ordering per fetcher in this particular crate since that'd require IO. I think smarter ordering would be the domain of e.g. sparkle.